lychee-lib/src/extract/html5gum.rs

```diff
@@ -0,0 +1,206 @@
+            | ("applet", "codebase")
+            | ("body", "background")
+            | ("button", "formaction")
+            | ("command", "icon")
+            | ("form", "action")
+            | ("frame", "longdesc")
+            | ("head", "profile")
+            | ("html", "manifest")
+            | ("iframe", "longdesc")
+            | ("img", "longdesc")
+            | ("input", "formaction")
+            | ("object", "classid")
+            | ("object", "codebase")
+            | ("object", "data")
+            | ("video", "poster") => {
+                Some(vec![attr_value].into_iter())
+            }
+            (_, "srcset") => {
+                let mut urls = Vec::new();
+                for image_candidate_string in attr_value.trim().split(',') {
+                    for part in image_candidate_string.split_ascii_whitespace() {
+                        if part.is_empty() {
+                            continue;
+                        }
+                        urls.push(part);
+                        break;
+                    }
+                }
+                Some(urls.into_iter())
+            }
+            _ => None,
+        }
+    }
+
+    fn flush_current_characters(&mut self) {
+        // safety: since we feed html5gum tokenizer with a &str, this must be a &str as well.
+        let raw = unsafe { from_utf8_unchecked(&self.current_string) };
+        self.links.extend(extract_plaintext(raw));
+        self.current_string.clear();
+    }
+
+    fn flush_old_attribute(&mut self) {
+        {
+            // safety: since we feed html5gum tokenizer with a &str, this must be a &str as well.
+            let name = unsafe { from_utf8_unchecked(&self.current_tag_name) };
+            let attr = unsafe { from_utf8_unchecked(&self.current_attribute_name) };
+            let value = unsafe { from_utf8_unchecked(&self.current_attribute_value) };
+
+            let urls = LinkExtractor::extract_urls_from_elem_attr(attr, name, value);
+
+            let new_urls = match urls {
+                None => extract_plaintext(value),
+                Some(urls) => urls
+                    .into_iter()
+                    .map(|url| RawUri {
+                        text: url.to_string(),
+                        element: Some(name.to_string()),
+                        attribute: Some(attr.to_string()),
+                    })
+                    .collect::<Vec<_>>(),
+            };
+
+            self.links.extend(new_urls);
+        }
+
+        self.current_attribute_name.clear();
+        self.current_attribute_value.clear();
+    }
+}
+
+impl Emitter for &mut LinkExtractor {
+    type Token = ();
+
+    fn set_last_start_tag(&mut self, last_start_tag: Option<&[u8]>) {
+        self.last_start_tag.clear();
+        self.last_start_tag
+            .extend(last_start_tag.unwrap_or_default());
+    }
+
+    fn emit_eof(&mut self) {
+        self.flush_current_characters();
+    }
+    fn emit_error(&mut self, _: Error) {}
```

Comment:
The `Emitter` trait could make most (all?) methods [optional](https://stackoverflow.com/a/51915705) and we could remove the empty ones here. There's probably a reason why this isn't the case, but I fail to see it.

Reply:
it's like the debate about whether a `match` stmt should be exhaustive or have a `_ => ()` case

in this case i feel like forgetting to impl one of these methods comes with risk of hard to spot bugs, as they all operate on shared mutable state (self) -- so I'd rather have every added method be a breaking change that the user needs to think about

Reply:
Fair enough.

Reply:
`emit_error` _could_ be safe to make defaulted but it depends on the implementation